### ISSUE TITLE
Typing tensorflow.data.Dataset

### ISSUE THREAD

**Issue by srcolinas** [tensorflow/tensorflow#53554]
**System information**
- TensorFlow version (you are using): 2.5
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

I have some functions that accept instances of tensorflow.data.Dataset. I would like to type those functions so that the caller knows what is expected. However, doing `data: tensorflow.data.Dataset`, is not good enough, because people wouldn't know what kind of structure is produced by the object (shape of tensors, whether is tuple or dictionary, etc).

There is another object, `tensorflow.data.DatasetSpec`, which can be used to document the objects that will be produces by the tensorflow.data.Dataset. However, I am not sure how to put those things together for a typing annotation, and I am left to describe the structure in docstring, which I do not think is ideal.

I would love to see something like a generic, where you pass the type it accepts, like `Dataset[Spec]`(like in lists we do `List[int]`), but I am happy to learn some other ways to achieve more descriptive typing than just `data: tensorflow.data.Dataset`.

**Will this change the current api? How?**
Only for typing purposes

**Who will benefit with this feature?**
A lot of people from the Machine Learning community using Tensorflow to build production systems. 

**Any Other info.**
I could help if provided with guidance.

**Comment by sachinprasadhs** [comment]
Could you provide some more details about this feature request with example use case.

**Comment by srcolinas** [comment]
Thanks @sachinprasadhs . The use case is to have strong type guarantees when used with static type checkes like `mypy`. Such type annotations and type checking is very useful for maintaining production code.

**Comment by aaudiber** [comment]
ccing @mdanatg who is more familiar with typing in TF

**Comment by mdanatg** [comment]
I think this would be blocked by `Spec` (as in `Dataset[Spec]`) being itself a type, which in turn is gated by Tensor being a parametric type, so you can write things like `Dataset[Tuple[Tensor[Int32], etc.]]` (see https://github.com/tensorflow/tensorflow/issues/12345). There has been an effort to do that, and got blocked by making `Tensor` a subtype of `typing.Generic` - a PR actually got submitted (https://github.com/tensorflow/tensorflow/pull/40921), but unfortunately some CPython-specific optimizations of EagerTensor broke and forced us to roll it back.

Long story short, this would become one step closer to reality once the PR can be resubmitted so that `Tensor` is a `Generic`.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
